### ISSUE TITLE
mesolve allows different dimensions for H and rho0

### ISSUE THREAD

**Issue by amcdawes** [qutip/qutip#479]
I'm still fairly new to QuTiP but I've found that `mesolve` allows `H` and `rho0` to have different dimensions. Of course getting it right is the users responsibility, an informative warning may be useful. A minimal example is as follows:
```
from qutip import *
import numpy as np

N = 3
a = tensor(destroy(N),identity(2))
sz = tensor(identity(N),sigmaz())

H = sz*a.dag() + sz.dag()*a
collapse = [a]
tlist = np.linspace(0,1,100)
solution = mesolve(H, basis(2,0), tlist, collapse, [a.dag()*a])
```
which gives a numerical integration error:
`UserWarning: zvode: Repeated convergence failures. (Perhaps bad Jacobian supplied or wrong choice of MF or tolerances.)`
Rather than a more direct warning about the (more likely) problem of mis-matched Hilbert spaces. I'm happy to add a check to the `mesolve` function and submit a PR, but I wanted to be sure that I'm not missing some use case where `H` and `rho0` should be allowed to differ.

**Comment by ajgpitch** [comment]
An explicit warning / exception sounds good to me.
You mention a UserWarning there as an error. Does that mean that the ODE solver raises an exception?  If so, then there can be no useful application of this, and hence we should raise an exception - probably a RuntimeError unless anyone has a better suggestion.

**Comment by nonhermitian** [comment]
I am surprised that this was not already checked. Anyway, it is really a good idea to has this implemented. One should be careful though since the input to mesolve can take a couple of forms, i.e one can input a Hamiltonian or Liouvillian.

**Comment by nonhermitian** [comment]
The input can also be a function that would be a bit more tricky to handle given the need to call the function first.

**Comment by amcdawes** [comment]
@ajgpitch yes, the solver does raise an error... but not always. I was able to get it to integrate a few times but it was very inconsistent. The other error it sometimes raises is "too much accuracy was requested for the precision of the machine". When I realized the issue, I was very surprised it even got that far and I still don't know how it could ever run the calculation with a bad `rho0`.

@nonhermitian yes, I see the point in the code where `mesolve` digs down into various types of `H` input. I'll work on something for the easy Qobj cases first, and think about how to handle the function case. Should be able to test the size of what the function returns.
